controllers/apps/apimanager_managed_routes_handler.go

```diff
@@ -0,0 +1,137 @@
+	if apimanagerReconcileRequest != nil {
+		q.Add(*apimanagerReconcileRequest)
+	}
+}
+
+// 	// Update is called in response to an update event -  e.g. Pod Updated.
+func (h *APIManagerRoutesHandler) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	apimanagerReconcileRequestOld := h.getAPIManagerOwnerReconcileRequest(evt.MetaOld)
+	if apimanagerReconcileRequestOld != nil {
+		q.Add(*apimanagerReconcileRequestOld)
+	}
+	apimanagerReconcileRequestNew := h.getAPIManagerOwnerReconcileRequest(evt.MetaNew)
+	if apimanagerReconcileRequestNew != nil {
+		q.Add(*apimanagerReconcileRequestNew)
+	}
+}
+
+// 	// Delete is called in response to a delete event - e.g. Pod Deleted.
+func (h *APIManagerRoutesHandler) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	apimanagerReconcileRequest := h.getAPIManagerOwnerReconcileRequest(evt.Meta)
+	if apimanagerReconcileRequest != nil {
+		q.Add(*apimanagerReconcileRequest)
+	}
+}
+
+// 	// Generic is called in response to an event of an unknown type or a synthetic event triggered as a cron or
+// 	// external trigger request - e.g. reconcile Autoscaling, or a Webhook.
+func (h *APIManagerRoutesHandler) Generic(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+	apimanagerReconcileRequest := h.getAPIManagerOwnerReconcileRequest(evt.Meta)
+	if apimanagerReconcileRequest != nil {
+		q.Add(*apimanagerReconcileRequest)
+	}
+}
+
+func (h *APIManagerRoutesHandler) getAPIManagerOwnerReconcileRequest(object metav1.Object) *reconcile.Request {
+	if object == nil {
+		return nil
+	}
+	h.logger.V(2).Info("Processing meta object", "Name", object.GetName(), "Namespace", object.GetNamespace())
+
+	for _, ref := range object.GetOwnerReferences() {
+		refGV, err := schema.ParseGroupVersion(ref.APIVersion)
+		if err != nil {
+			h.logger.Error(err, "Could not parse OwnerReference APIVersion",
+				"api version", ref.APIVersion)
+			return nil
+		}
+
+		h.logger.V(2).Info("Evaluating OwnerReference", "GroupVersion", refGV, "Kind", ref.Kind, "Name", ref.Name)
+		// Compare the OwnerReference Group and Kind against the APIManager Kind
+		// or the name of the Zync Que DeploymentConfig.
+
+		// If the OwnerReference of the received object is an APIManager we
+		// return a reconcile Request using the name from the OwnerReference and the
+		// Namespace from the object in the event
+		apimanagerKind := appscommon.APIManagerKind
+		apimanagerGroup := appsv1alpha1.GroupVersion.Group
+		if ref.Kind == apimanagerKind && refGV.Group == apimanagerGroup {
+			h.logger.V(2).Info("APIManager OwnerReference detected. Reenqueuing as APIManager event", "APIManager name", ref.Name, "APIManager namespace", object.GetNamespace())
+			// Match found - add a Request for the object referred to in the OwnerReference
+			request := &reconcile.Request{NamespacedName: types.NamespacedName{
+				Name:      ref.Name,
+				Namespace: object.GetNamespace(),
```

Comment:
I know the operator is namespaced and this code works because it assumes namespacing. Shouldn't be better to make it more generic and make sure the route's namespace is the same as the APIManager CR namespace?

Reply:
OwnerReferences cannot have references to objects on other namespaces so this should never be possible. See:
https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference

This means that if the route contains an OwnerReference to an APIManager, the APIManager will be in the same namespace as the route

Reply:
Good to know